### PR TITLE
Update consul-template to v0.25.1

### PR DESCRIPTION
**Pull request by fredrikhgrelland**
While working on https://github.com/hashicorp/nomad/pull/8849 I needed consul namespace support in consul-template.

Inclusive language changes required some refactoring, but is kept backwards compatible.

### CODE REVIEW

File: command/agent/config.go [hunk 1]
```diff
@@ -307,9 +307,10 @@ type ClientConfig struct {
 // rendering
 type ClientTemplateConfig struct {
 
-	// FunctionBlacklist disables functions in consul-template that
+	// FunctionDenylist disables functions in consul-template that
 	// are unsafe because they expose information from the client host.
-	FunctionBlacklist []string `hcl:"function_blacklist"`
+	// COMPAT(0.13) consul-template uses inclusive language from v0.25.0 - function_blacklist is kept for compatibility
+	FunctionDenylist []string `hcl:"function_denylist" hcl:"function_blacklist"`
```
Comment (tgross) [inline command/agent/config.go:313]: I built Nomad from this PR. The `function_denylist` configuration does work to block the function.

But to verify backwards compatibility I tried to use the older `function_blacklist` syntax in the configuration: it's accepted by the client config parsing but it doesn't end up blocking access to the function anymore. I'm not certain this syntax for go field tags actually works? A couple things you could try here:

* Try with the fields tags swapped
* Add a new field for `function_denylist` and have it accept the value of `function_blacklist` if not set.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,7 @@ IMPROVEMENTS:
  * core: Improved job deregistration error logging. [[GH-8745](https://github.com/hashicorp/nomad/issues/8745)]
  * api: Added support for cancellation contexts to HTTP API. [[GH-8836](https://github.com/hashicorp/nomad/issues/8836)]
  * driver/docker: upgrade pause container and detect architecture [[GH-8957](https://github.com/hashicorp/nomad/pull/8957)]
+ * client: consul-template updated to v0.25.0 [[GH-8988](https://github.com/hashicorp/nomad/pull/8988)]
```
Comment (tgross) [inline CHANGELOG.md:7]: Can you call out the client config deprecated here as well?

There's currently a merge conflict on this CHANGELOG file. Also (and this probably seems silly-picky) we sort the changelog entries by `core:`, then other sections alphabetically, and then by line length. So if you could move this to just above the `driver/docker` entry that'd be great.

**Review by tgross — CHANGES_REQUESTED** [review]
Hi @fredrikhgrelland! Thanks so much for this PR, especially for pulling out the consul-template change from the large PR you're working on.

---

This isn't a blocker but would help testing: to test backwards compatibility, I ran a job on a build from `master`, then built this PR and restarted Nomad. Unfortunately I got a panic when restarting Nomad with the new build (see below). If I wipe out the data dir and run with this build, it works fine. And when I rebased this PR on master the problem and started over from scratch the problem went away... there's probably something that landed on master that you don't have in your branch that's breaking raft entries, so if you were to rebase from master that would make it easier to test this out. Thanks!

```
Started Nomad.
==> WARNING: Bootstrap mode enabled! Potentially unsafe operation.
==> Loaded configuration from /etc/nomad.d/standalone.hcl
==> Starting Nomad agent...
panic: failed to apply request: []byte{0x48, 0x81, 0xaa, 0x43, 0x72, 0x65, 0x61, 0x74, 0x65, 0x54, 0x69, 0x6d, 0x65, 0xd3, 0x16, 0x39, 0xdd, 0xcf, 0x8d, 0x5b, 0xd7, 0xf9}
goroutine 16 [running]:
github.com/hashicorp/nomad/nomad.(*nomadFSM).Apply(0xc0002d73b0, 0xc0005d0aa0, 0x4f8ad20, 0xbfd590b12b50cc03)
        github.com/hashicorp/nomad/nomad/fsm.go:292 +0x193b
github.com/hashicorp/raft.(*Raft).runFSM.func1(0xc00007cac0)
        github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/fsm.go:90 +0x2c1
github.com/hashicorp/raft.(*Raft).runFSM.func2(0xc000212000, 0x33, 0x40)
        github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/fsm.go:113 +0x75
github.com/hashicorp/raft.(*Raft).runFSM(0xc0005f4000)
        github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/fsm.go:219 +0x42f
github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc0005f4000, 0xc0003d6c00)
        github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/state.go:146 +0x55
created by github.com/hashicorp/raft.(*raftState).goFunc
        github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/state.go:144 +0x66
```

File: command/agent/config.go (hunk 1 above)
Comment (tgross) [reply command/agent/config.go:313]: Ok, it looks like the `template` stanza doesn't validate its keys fully (I saw this on another sub-section recently so I think it's a general bug in our config parser), so it'll accept nonsense fields. For example, this client block works just fine:

```hcl
client {
  enabled = true

  template {
    whatevs = true
    function_denylist = ["env"]
  }
}
```

So in this branch, the `function_blacklist` is being accepted by an overly generous config parser.

Comment (tgross) [reply command/agent/config.go:313]: I've opened https://github.com/hashicorp/nomad/issues/9001 about the overly-generous config parser.

Comment (fredrikhgrelland) [reply command/agent/config.go]: This was very interesting. It worked if I left it as default (["plugin"]), but it did NOT fail if `function_blacklist = ["plugin"]`. It did however work to disable the denylist with both `function_blacklist = [""]` and function_denylist = [""]

Either way, changed it so both variables are available.

test job:
```
job "test" {
  datacenters = ["dc1"]
  group "test" {
    task "example" {
      driver = "raw_exec"

      config {
        command = "/bin/sleep"
        args    = ["100"]
      }
      template {
        data        = "{{ plugin \"cat\" \"/etc/hosts\" }}"
        destination = "local/file.txt"
      }
    }
  }
}
```

👍 task **fails** without any config

👍 task **fails** with config ( equals default )
```
client {
  enabled = true
  template {
    function_denylist = ["plugin"]
  }
}
```

👍 task **fails** with config ( equals deafult )
```
client {
  enabled = true
  template {
    function_blacklist = ["plugin"]
  }
}
```
👍 task **works** with config
```
client {
  enabled = true
  template {
    function_blacklist = [""]
  }
}
```
👍 task **works** with config
```
client {
  enabled = true
  template {
    function_denylist = [""]
  }
}
```

File: CHANGELOG.md (hunk 2 above)
Comment (fredrikhgrelland) [reply CHANGELOG.md]: Done.

**Comment by fredrikhgrelland** [comment]
Rebased from master and ready for review.

**Review by tgross — APPROVED** [review]
I've checked out the rebase PR and tested it out and everything looks great! :+1:

**Pull request merged**